compiler/modules/front/expr.bal

```diff
@@ -1158,6 +1160,26 @@ function validArgumentCount(ExprContext cx, bir:FunctionRef func, int nSuppliedA
     }
 }
 
+function validArgumentTypes(ExprContext cx, bir:FunctionRef func, bir:Operand[] suppliedArgs, s:MethodCallExpr|s:FunctionCallExpr expr) returns CodeGenError? {
+    foreach int i in 0 ..< suppliedArgs.length() {
+        if operandHasType(cx.mod.tc, suppliedArgs[i], func.signature.paramTypes[i]) {
+            continue;
+        }
+        s:Expr argExpr;
+        if expr is s:FunctionCallExpr {
+            argExpr = expr.args[i];
+        }
+        else if i == 0 {
+            argExpr = expr;
+        }
+        else {
+            argExpr = expr.args[i - 1];
+        }
+        return cx.semanticErr(`wrong argument type for parameter ${i + 1} in call to function ${symbolToString(cx.mod, cx.defn.part.partIndex, func.symbol)}`, s:range(argExpr));
```

Comment:
A better way to solve this problem is to add a `codeGenExprForType` function (like codeGenExprForBoolean/String), which takes a semType argument. Then use that instead of codeGenExpr in codeGenMethodCallExpr/codeGenFunctionCallExpr.

Add test cases which have arguments on different lines to check that the error is for the right line.

For the too many arguments case, the error message should have the location of the first excess argument.

For the too few arguments case, the error message should be the location of closing parenthesis of the function call (may need to fix the AST to do this.)

There's no need for the error message to mention the parameter number or function name give that we have ranges. This applies for both wrong number and wrong type of messages.